### DISCUSSION TITLE
Advanced Triggers

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#527] [REM]
Can you provide some examples on how to use "Binding" and "Filter Expression"?

I'm trying to implement boolean login in REM for Multi-Relay Boards.

**Comment by orrious** [comment]
Scenario 1:  When Pool/Spa virtual circuit is activated, I would like to turn on multiple "relays".  Currently if I define two triggers, one for each relay, with the same "Basic Filter", both relays turn on, but then one shuts off immediately.

What is the proper way to turn both relays on at the same time when Pool/Spa is enabled?

**Comment by orrious** [comment]
Scenario 2: IF statements.  IF (virtualcircuit: "Pool/Spa" == on) or (circuit: "Spillway == on ); then on

**Reply by rstrouse** [reply]
The advanced binding and filters for triggers are closures.  Inside of these closures you can use javascript to transform the incoming data.  In the execution context you will find the following variables.
```
connection:  This is the monitored connection that the trigger is coming in from.
trigger:     The trigger definition that fired the trigger.
device:      The definition of the device that the trigger belongs to.
data:        Raw incoming object data
```

However, if you give me a bit of background as to what you are trying to do, perhaps I can formulate a trigger for you that would get you going.

**Reply by orrious** [reply]
The spa has a drain and two returns; one in the floor for normal operation and the other the spa jets.  This would normally be a typical setup, however the spillway is large between the pool and spa.  The water just trickled out of the spa.

I have waterfalls on a single speed pump.  They have a "relief return" valve that dumped into the pool to bypass the extra pressure from the waterfalls.  

When I replumbed my pool, I wanted to harness that extra flow when the waterfalls were on.  So with 2 check valves and another intellivalve, I plumbed it so the "relief return" could be switched from the Spa Jets to the pool.  Unless it's in spa mode, the spa jets fixtures aren't used, so I used them for the "relief return".  The default position is for the waterfalls to use the spa jets as the return.  Now when I turn on the waterfalls, the spillway really flows!

Also the spa return is also on an intelivalve and is what is used to turn off the Spillway in normal pool mode.

I would want the relay for the valve "FeatureReturn" to be activated:

IF ((circuit: "Spa" == on) and (feature: "Waterfalls" == on)) or ((circuit: "Spillway" == on) and (feature: "Waterfalls" == on))

**Reply by rstrouse** [reply]
If you simply trap the valve socket from njsPC it should trigger such a valve in REM.  njsPC should be able to activate this from a logical state.

**Reply by rstrouse** [reply]
Hey.  Imagine for a second if you could define an interface file that would allow you to hook events in njsPC.  In that instance you would be able to modify behavior and generate your own socket output to REM.  I have been itching to add the processor functions to the socket outputs.

**Reply by orrious** [reply]
All depends on where you want to implement the logic.  I run in Nixie mode, so I don't care about backward compatibility with other controllers. If we modify logic in njsPC, would that only be applicable in Nixie mode?  That said, we need a place for advanced logic, whether it be njsPC, REM, or even externally like NodeRed, etc.

Maybe this is a stretch, is there any chance NodeRed can be leveraged to give us a view of what's going on inside of njsPC and use it to modify it?

**Reply by rstrouse** [reply]
It would end up transparent and local to the specific implementation.  All other interfaces and OCPs would be none-the-wiser.  This type of thing is very specific to the construction of the pool.  Something very similar was recently done to allow a Nixie controller to speak to Aqualinkd which was in turn speaking to Jandy iAqualink.

The beauty is that if you never installed the interface it had no impact on you.  Further, if you had any other interface it just continued to operate as the interface was just enforcing additional rules that were required for the iAqualink OCP.

Yes NodeRed can be used to modify logic.  But why add another layer.  Most of the time all that is needed are a few scripts to color outside the lines.  If it is maintained at the njsPC level it has access to the current state for all equipment attached to the pool.

**Reply by orrious** [reply]
I was hoping it wasn't another layer per se, but I guess it could end up that way.  I was thinking of using the XML style from NodeRed so you could import/export from one to another.  For me, when troubleshooting or modifying the logic, NodeRed is very easy.  You don't need to know any language to use it and it's very visual and easy to troubleshoot, etc. 

Is there any way you could leverage NodeRed to display the current logic visually or something similarly?  I don't want to reinvent the wheel

**Reply by rstrouse** [reply]
Yeah not really interested in the weight of embedded node-red.  Scripting on events is really lightweight and provides a whole lot of power to extend existing functionality with very little cost and no cost when additional functionality is not needed.  All of this would be in standard javascript with access to the global state context and no impact on updates.  I couldn't imagine building flows for all the supported equipment options in njsPC.

**Reply by orrious** [reply]
Makes sense, I wasn't trying to embed node-red in either.  I wouldn't want you to build flows for each.  I was overly optimistic it might have just plugged in.  However you go about it, keep in mind a docker deployment, persisting the "rule changes" between versions, etc.

**Reply by orrious** [reply]
I haven't had a need until now to learn javascript.  Is REM easier to start digging through to learn?

**Reply by rstrouse** [reply]
Perhaps I can equate it to any languages you currently have experience with.  It is probably the easiest language to master.  And in the context we are referring to here it does not require any compilers, linkers, or interpreters.   In fact after I extend the interface definitions changes can be made in real time.  It will subscribe to the internal events and provide a simple place to add snippets.

For instance your pseudo code above could be written as follows and would be triggered any time the state of the intake/return valves change or waterfall state changes.  Setting the state for the featureReturn would trigger all things in njsPC affiliated with it to include any pump changes, valve changes, circuit group changes, and REM calls.

```
let waterfall = state.features.getItemById(129); // 128 is the id of the waterfalls feature 
let valve = (state.valveMode.name === 'spillway' || state.valveMode.name === 'spa');  // If the valves are in spillway or spa then true
if(valve && waterfall.isOn)
   sys.board.features.setFeatureState(130, true);  // Turn on hidden feature for featureReturn
else
   sys.board.features.setFeatureState(130, false);  // Turn off hidden feature featureReturn
```

**Reply by orrious** [reply]
ok,  Appears to be similar to python in that it's object oriented and interpreted rather than compiled.  I need to look at the structure but won't take me long to get up to speed.  Take it you can load libraries, etc.  Time to find a 'hello world' for js

**Reply by tagyoureit** [reply]
> Hey. Imagine for a second if you could define an interface file that would allow you to hook events in njsPC. In that instance you would be able to modify behavior and generate your own socket output to REM. I have been itching to add the processor functions to the socket outputs.

@rstrouse Glad to see you are finally understanding what I was trying to do with rulesManage.json ~2 years ago!  Lol.  I'm sure yours will be much more capable.  ;)

**Reply by rstrouse** [reply]
@tagyoureit yes I have garnered a better understanding.   @orrious if you upload a backup of your system I will build the interface that you are looking with all the proper ids.

**Reply by orrious** [reply]
Here is my backup, less personal info.

[backup.zip](https://github.com/tagyoureit/nodejs-poolController/files/9128635/backup.zip)

**Reply by rstrouse** [reply]
Ok so pull njsPC and dashPanel.  I added the rule interface.  This allows you to catch events in njsPC and react to them with user defined rules.

There are a few things that you will need to do to enable the functionality as I understand it.

Step 1:  Add a Feature from the dashPanel configuration pages and name it whatever you like.  When njsPC creates this feature it will stamp it with id 129 since this is the first one.
Step 2: Assign your vFeatureReturn valve to this feature under the valves tab in dashPanel.  If you uncheck Show as Feature it will hide this in the dashPanel display.
Step 3: Save the JSON below to a file.  The name is only important to you so call it what you like.  I called mine RuleProcessor.json.
```
{
  "context": {
    "name": "ruleProcessor",
    "options": {},
    "vars": {}
  },
  "events": [
    {
      "name": "valve",
      "description": "Notify when either the intake or return valve is triggered",
      "processor": [
        "let waterfall = state.circuits.getItemById(3);  // From poolConfig.json the id of waterfalls is 3.",
        "let valve = state.valveMode >= 2;  // 0=off, 1=pool, 2=spa, 3=spillway",
        "sys.board.features.setFeatureStateAsync(129, valve && waterfall.isOn); // Turn on the feature that controls the waterfall valve"
      ],
      "enabled": true,
      "filter": "@bind=data.isIntake; || @bind=data.isReturn;"
    },
    {
      "name": "circuit",
      "description": "Notify when the waterfall feature gets turned on",
      "enabled": true,
      "processor": [
        "let waterfall = state.circuits.getItemById(3);  // From poolConfig.json the id of waterfalls is 3.",
        "let valve = state.valveMode >= 2;  // 0=off, 1=pool, 2=spa, 3=spillway",
        "sys.board.features.setFeatureStateAsync(129, valve && waterfall.isOn); // Turn on the feature that controls the waterfall valve"
      ],
      "enabled": true,
      "filter": "@bind=data.id; === 3;"
    }
  ]
}
```
Step 4: Under the configuration => Controller => Interfaces tab click Add Interface and upload the bindings file that you just saved by clicking on the Upload Bindings button, selecting the file, and pressing the Upload File button.  This places the file into the njsPC custom bindings directory `nodejs-poolController/web/bindings/custom`.
Step 5:  Give your interface a name, select the type as Rule and choose the binding from the binding dropdown.  Then press the Add Interface button.
![image](https://user-images.githubusercontent.com/47839015/179551879-acf2f71a-865f-49b1-bfae-e671d2221ff3.png)
Step 6: This above action will add an interface to your interfaces list with the name you gave it.  Expand the interface, check the Enabled checkbox and press Save Interface.  The rules will now be applied from the bindings file you assigned to the interface.

There is an incredible amount of power available in the rule processor.  I think I captured the functionality that is unique to your pool and will be happy to explain anything that seems unclear.

The following objects are available inside the processor.
```
data: // This is the payload for the event and matches all the standard interface data structures for the incoming event.  See the Data Structures wiki for a list.
sys: // The system configuration object including the board controller.  This allows setting other circuits, features, and equipment.
state: // The current state of all items on the controller
logger: // Allows you to write to the log with logger.info(`message`), logger.warn(`message`), logger.error(`message`)... See Logger.ts
webApp: // Send socket data emits and access other interfaces.
```

NOTE:  Because I know you will ask the filter is a Boolean expression that is evaluated to route the events.  For instance, in the above interface bindings, whenever the valve event is triggered, the preprocessor looks at the incoming data payload to determine if the valve event is for either the intake or return.  If it is not then it does not execute the defined processor.  It goes like this the tokenizer looks to any tokens that are in the form `@bind=<value reference>;` and replaces this token with the actual value.  So when a valve state changes it looks at the incoming event data for the valve and evaluates the expression.  If you look at poolState.json you will see the valve state that is emitted for the intake and return valves.

**Reply by orrious** [reply]
Followed all steps above.  But I think I'm lacking context on a couple key items.  

Circuit IDs:  How are they set?  Where do you find them?

Intake and Return valves:  I deleted these originally and they came back.  I don't currently leverage them as their names were unchangeable and meanings not relevant.  I'm getting the feeling these are "special".  In a Shared Body which valves are these supposed to be?

**Reply by orrious** [reply]
Ok, it helps to RTM.   https://github.com/tagyoureit/nodejs-poolController/wiki/Nixie-Controller-Setup shows where Intake and Return should go.  I've made the adjustments and delete my vSpaDrain and vSpaJets valves, as they are replaced by intake and return, respectively.

**Reply by rstrouse** [reply]
You will find the circuit and feature ids in the poolConfig.json or the poolState.json.  There will be 1 entry in each file for each circuit, feature, or equipment item.

**Reply by orrious** [reply]
When Cleaner circuit is activated via schedule 8am-11am: 
IntelliChlor doesn't run, I would like it to.
Filter show pressure, but doesn't show as being active

Turning the "Pool" on does not turn the "Cleaner" off, but does make Filter and IntelliChlor active, I had it working on my own where it worked like Pool / Spa, you turn one on and the other turns off automatically.

**Reply by orrious** [reply]
Re-reading this, it came off short which certainly wasn't my intentions.  Want to make sure I don't come off ungrateful.  I really appreciate all the effort y'all put into this project.

**Reply by tagyoureit** [reply]
All good!

I see you have the cleaner circuit set as a 'generic' type which means it isn't tied at all to the main Pool circuit.  The intellichlor will only run when the main pool (or spa) circuit is on.  If you set the cleaner circuit to be of 'master cleaner` type you will find it is tied to the pool circuit and cannot be turned on unless the pool circuit is on.  Even with a schedule it would need to be on only when the pool circuit is manually turned on or is on from a schedule.

<img width="620" alt="2022-07-19_14-41-00" src="https://user-images.githubusercontent.com/1646840/179854154-62d40169-4840-4b36-942e-b09934a85933.png">

**Reply by orrious** [reply]
@rstrouse said earlier this is not what I wanted.  My cleaner runs at a lower speed than the pool, so the pool circuit overrides it.

**Reply by tagyoureit** [reply]
Your pool circuit should be the lowest speed and then use additional features/circuits to increase the speed.  We won't turn on the chlorinator without the pool circuit being active. 

Hence, your cleaner won't need to increase the pump speed but you can use additional features/circuits to increase speed when the cleaner is not active.

**Reply by orrious** [reply]
If I have to create another feature just to run the pool at the correct speed, it's going to make using Dash Panel awkward.  The big Pool button will loose it's function all together.  Is there a way to create this functionality in the ruleProcessor?

**Reply by rstrouse** [reply]
The pool circuit should be scheduled to run such that it encompasses your cleaner circuit.  The cleaner circuit should not be set to cleaner in your case since it only rotates a valve and does not engage additional booster pumps.   Then you create your cleaner schedule so that it runs inside the pool schedule.

Your pump configuration should include the pool circuit to run at the lowest speed available.  Then when the cleaner circuit kicks on it will override the pool speed to run your cleaner when the cleaner is finished it will resume the pool speed.  Pump speeds ramp to the highest needed rpm to run the circuit.

**Reply by orrious** [reply]
If the cleaner speed is the lowest: 2000 and I need 2500 for the pool.


Pool to 2000
Cleaner to 2000
Wouldn't I need a Feature or new Circuit to run the pool at 2500?

**Reply by rstrouse** [reply]
Typically the base pool operation should be the lowest speed available to engage the IntelliChlor flow sensor + ~200rpm to account for filter cleanliness.  

Then as you turn on other circuits and features these most of these will require additional flow.  For instance rotating the cleaner valve will most likely divert flow or suction depending on whether the cleaner is a suction side or pressure side cleaner.  If it also has a booster pump then you must increase the rpm above the baseline because cavitating such a pump would be bad.  They are high flow pumps that cannot manage suction.

If both of them only require 2000 then you can set both to 2000.   If you want additional pump volume when you turn on the spillway or spa simply assign them as pump circuits as well.  The idea is that when you run your pool at the lowest rpm required to accomplish the task it will save on energy.  

Ultimately, though the chlorinator should never run unless the designated body circuit is engaged.  In fact you should tie its power into the filter circuit to make sure it doesn't.   The flow switch on the chlorinator should be a last resort.  

I have one on my bench that failed in the on position and had been running like that for about a year.  Had I been relying on the flow switch it could have amounted to hydrogen gas buildup in the highest points of the plumbing.  There are lots of stories of exploded pads out there and it is the reason that the wiring instructions insist on it being tied to the filter relay.

**Reply by tagyoureit** [reply]
Just curious, too... why do you run your pool at such a high speed?  I typically run ~1500rpm for my baseline, and ~1800rpm for solar, 2300rpm when heating with my heater and even higher for the spa jets.  I know everyone's pool is different and maybe it's just for aesthetics, but my pool (and energy bills) are really cranking at the 2500+rpm level.

**Reply by orrious** [reply]
Chlorinator will be on it's own circuit, as soon as the SSR comes in.  The filter pump / main pump is VSF and is only on a breaker.

How do you set the Spillway up so that it is on all the time, unless its turned off with an egg timer that will turn it back on?  Or do you invert it?

**Reply by tagyoureit** [reply]
<img width="624" alt="2022-07-19_17-52-46" src="https://user-images.githubusercontent.com/1646840/179872489-b4eaef7f-d98f-48e3-a228-8b21790d7a4a.png">

The pump circuits @rstrouse are mentioning are here... you can set speeds for the solar, heater, etc without needing to define additional circuits for them.

> Chlorinator will be on it's own circuit, as soon as the SSR comes in.

NO NO NO.  This needs to be on the same breaker as the main filter pump.  This is what @rstrouse is saying above.  It's a safety hazard if they aren't tied together.

**Reply by orrious** [reply]
It's on the same breaker.  The pump is not on a SSR, it's controlled by RS485.  So I need a SSR for the Chlorinator.  In njsPC I will need a Chlorinator circuit and tie it to Poo/Spa, so it comes on any time the Pool or Spa is on.

What am I missing here?

**Reply by orrious** [reply]
For the Spillway.  When it has a green light on the Spillway feature, is that to indicate that the Spillway is active or inactive?  i.e should water be flowing over the spillway when the green light is on or off?

**Reply by rstrouse** [reply]
Typically in a pool automation system these use 3hp relays that have dual contacts on the other side of the trigger.  With SSRs the equivalent would be to gang the trigger or wire the transformer to the load side of the filter relay.  
https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-Standalone-Nixie-Pool-Controller#high-current-circuits

A green light indicates that the circuit or feature is on.  Water should be flowing over the spillway when the green light is on.

**Reply by tagyoureit** [reply]
[EDIT - Dang, rstrouse beat me this time.]

Chlorinator - If it were me, and as also recommended by the manuals for the OCP's, I'd setup the SSR so it turns on when the pool circuit is on.  Then you connect the chlorinator to the filter pump relay.  The danger is if you turn on the chlorinator but the pump isn't running.  You don't need a separate chlorinator circuit to turn on.  I understand you want to set it up differently based on your cleaner, but it's a risk I wouldn't take.  See the section of the install manual below.  

<img width="981" alt="2022-07-19_18-44-01" src="https://user-images.githubusercontent.com/1646840/179877898-b03061fa-e628-482b-8882-ff10faa44a63.png">

For the spillway, green=on=spillway active.  If your valve is operating backwards from what you want you can flip the toggle switch and it will simply reverse the polarity and like Stranger Things it will go to the Upside Down.

**Reply by orrious** [reply]
I pieced my own components together.  There is no filter relay nor has there ever been one.  The closest thing I have to it currently is for my ozone generator, but it is on a dedicated breaker as well.

The only way for me to have the chlorinator on the same breaker as the pump and be controlled when the pump comes on/off would be what I've proposed.  Add another SSR between the pool pump breaker and the chlorinator.  Create a new valve, yes valve, in njsPC that comes on/off with Pool/Spa.  This is currently how my ozone generator is working.  

Am I missing something?

**Reply by rstrouse** [reply]
I know that it seems redundant but I would create that filter relay.   You assign this relay in Nixie under the filter configuration and it has a latch that drops it should any failure occur.   All the major manufacturers of pool automation to include Intermatic time clocks suggest that the filter circuit be tied to the SWG.

@tagyoureit I beat U again!

![image](https://user-images.githubusercontent.com/47839015/179880092-8e5a1a10-c66c-4e4a-a345-1b278af441d8.png)

**Reply by tagyoureit** [reply]
😢 You won this battle, but the war isn't over!

**Reply by orrious** [reply]
Ok, this makes much more sense.  You both kept mentioning a "Filter Circuit" or "Filter Relay"  I didn't realize there was literally a Circuit attached to the filter that had special logic behind it.

**Reply by rstrouse** [reply]
Sorry about that I thought you just were resisting implementing it.

**Reply by orrious** [reply]
Not at all.  I didn't put 2 & two together.  I hadn't seen this elusive Filter Relay you spoke of.  It's hidden away from everything that does "Circuit" type work.  Honestly the first value I had set there was the Pressure sensor.  I figured that's how it knew which pressure feed to use.

**Reply by orrious** [reply]
For Settings -> Bodies -> Filters -> Filter Pressure > Circuit, currently it is Pool.  Should this be Pool/Spa, if not, why not?

**Reply by rstrouse** [reply]
You typically want to measure the filter clean pressure when it is performing the most basic filtering function.  That’s is what the circuit selection is.  It will make sure it is the only thing running when it compares the filter percentage.  It ignores all circuit or feature not attached to a valve or a pump as well as lighting.

**Reply by orrious** [reply]
For the Rule Interface, does the enable button work?  I.e can you disable rule processing to validate no-rule behavior?

**Reply by orrious** [reply]
I currently have the Rule Interface disabled, hopefully that works and isn't causing the issues I'm having. 

Intake -> vSpaDrain, type Pool/Spa (currently when active enables the Spa Drain so the filter pump only gets Spa water)
Return -> vSpaJets, type Pool/Spa (currently when active, diverts all water returning to Spa Jets.)
vSpaReturn, type Spillway.  (currently when active, divers water from returning to spa, shuts spillway off, I need to invert this in REM)

When in Spa mode, Intake (vSpaDrain) and Return (vSpaJets) activate, as I would expect.
When in Pool mode and Spillway is ON:  Return (vSpaJets) and vSpaReturn are active.  Why is Return (vSpaJets) also activating and how do I stop it?

**Reply by rstrouse** [reply]
There are two states for valves.  These are diverted and undiverted.   The undiverted lead is typically connected to the Normally Closed (NC) lead for the valve.  If it needs to be diverted you can electrically change it either with the toggle switch on the valve (CV24) or by holding the left and right buttons (IntelliValve) for a few seconds.

When in pool mode the Intake and Return valve should be undiverted.  In spa mode both intake and return will be diverted and in spillway only the return valve is diverted.

All other valves are diverted or undiverted based upon the circuit defined.  If the circuit is on then the valve will be diverted.  What is vSpaReturn defined as.  By chance is it a Spillway type?  Valve mode Spillway is engaged whenever any circuit or feature is on that is defined as a spillway type.

**Reply by orrious** [reply]
I think this assumes that there is only 1 return to the spa, where I have 2, vSpaReturn (bottom of the spa) and vSpaJets (the jets).

How can we get the logic for the following?  Keep in mind the default state will be to have the Spillway ON, so default will be Diverted and will use inverted signal on REM.  Why?  No point in running a relay all the time and only turning it off when you want the Spillway off.

Pool ON / Spillway OFF:
vSpaDrain : OFF (not diverted)
vSpaJets : OFF (not diverted)
vSpaReturn : OFF (not diverted)

Pool ON / Spillway ON:
vSpaDrain : OFF (not diverted)
vSpaJets : OFF (not diverted)
vSpaReturn : ON (diverted)

Spa mode: ON
vSpaDrain : OFF (not diverted)
vSpaJets : OFF (not diverted)
vSpaReturn : ON (Preferred ON, but honestly doesn't matter as vSpaJets are the first valve in the plumbing and will override it.)  

Spa Drain: ON
vSpaDrain : ON (diverted)
vSpaJets : OFF (not diverted)
vSpaReturn : OFF (not diverted)

I'm just trying to get back functionality I lost when moving to REM

**Reply by orrious** [reply]
Yes vSpaReturn is type Spillway.  That way it is diverted when the Spillway is active.  I assume Spillway also diverts Return

**Reply by orrious** [reply]
Getting my head around this more.. 

1.) deleted the valve vSpaReturn
2.) changed the Return valve to vSpaReturn valve in REM
3.) created a new valve for vSpaJets with Circuit Spa and tied to REM vSpaJets

**Reply by rstrouse** [reply]
There are lots of options here because you can extend the valve operation with the ruleProcessor.
